lib/Doctrine/DBAL/Types/TypeRegistry.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Types;
+
+use Doctrine\DBAL\DBALException;
+use function array_search;
+use function get_class;
+
+/**
+ * @internal
+ */
+final class TypeRegistry
+{
+    /**
+     * Map of already instantiated type objects. One instance per type (flyweight).
+     *
+     * @var Type[]
+     */
+    private $typeInstances = [];
```

Comment:
This data structure doesn't prevent multiple keys pointing to the same instance, neither does the implementation of `addType()` (multiple names → same class). Would it make sense to use `SplObjectStorage` where instances were keys?

Reply:
This structure doesn't, but current API does. If we decide to change API, then we may need to change this too.

Reply:
`SplObjectStorage` may not be the best fit for this use case since the primary use case is _finding objects by keys_. IMO `array` is fine as long as we maintain consistency inside TypeRegistry and don't leak it outside.

Reply:
Agree with the above. I still don't see how `register()` prevents subsequent calls like `register('type1', T::class)`, `register('type2', T::class)`.

Reply:
Please see the new version (just pushed) where register() accepts instances directly.